utils/merge_syft_sbom.py

```diff
@@ -0,0 +1,128 @@
+#!/usr/bin/env python3
+import json
+import sys
+
+from typing import Any, Callable
+from urllib.parse import urlsplit, quote_plus
+
+
+# Functions to detect components that are reported differently by Syft
+def _is_syft_local_golang_component(component: dict) -> bool:
+    """
+    Check if a Syft Golang reported component is a local replacement.
+
+    Local replacements are reported in a very different way by Cachi2,
+    which is why the same reports by Syft should be removed.
+    """
+    return (
+        component.get("purl", "").startswith("pkg:golang")
+        and component.get("name", "").startswith(".")
+        or component.get("version", "") == "(devel)"
+    )
+
+
+def _is_cachi2_non_pypi_pip_dependency(component: dict) -> bool:
+    """
+    Check if Cachi2 Pip component was fetched a VCS or HTTPS location.
+
+    Cachi2 reports non-pypi components in a different way from Syft, so
+    the reports from Syft need to be removed.
+
+    Unfortunately, there's no way to determine which components are non-Pypi
+    by looking at the Syft report alone. This function is meant to create a
+    list of non-Pypi components from Cachi2's SBOM, then remove the ones reported
+    by Syft in the merged SBOM.
+    """
+    return (
+        component.get("purl", "").startswith("pkg:pypi")
+        and "version" not in component
+    )
+
+
+# Unique key functions
+def _unique_key_cachi2(component: dict) -> str:
+    """Create a unique key from Cachi2 reported components."""
+    url = urlsplit(component["purl"])
+    # omit the query part, since Cachi2 sets it and Syft does not
+    return url.scheme + ":" + url.path
+
+
+def _unique_key_syft(component: dict) -> str:
+    """Create a unique key for Syft reported components."""
+    # very few Syft components don't have a purl (e.g. type=OS)
+    if "purl" not in component:
+        return component.get('name', '') + '@' + component.get('version', '')
+
+    if "@" in component["purl"]:
+        parts = component["purl"].split("@")
+        # We need to encode special characters in the version part,
+        # and also need to ensure the name is all lowercase
+        return parts[0].lower() + "@" + quote_plus(parts[1])
+    else:
+        return component["purl"]
+
+
+def _get_syft_component_filter(cachi_sbom_components: list[dict[str, Any]]) -> Callable:
+    """
+    Get a filter function for Syft components in the SBOM.
+
+    This function applies several rules to determine if a component should be
+    removed from the final merged SBOM.
+    """
+    cachi2_non_pypi_components = [
+        component["name"] for component in cachi_sbom_components
+        if _is_cachi2_non_pypi_pip_dependency(component)
+    ]
```

Comment:
Is it safe to de-duplicate by `name` only? While it's not possible to have two different versions of the same dependency in one requirements file, it's possible to have them across different req files or across two different packages

Reply:
Yeah, I thought about that. Since Cachi2 does not report the version for vcs/file dependencies, there's no way to accurately determine that it is exactly the same dependency. 

This won't be a problem if all packages are processed by Cachi2, since we'll get Cachi2's report of all versions of this dependency. The corner case I see is if one of the packages is not processed by Cachi2, or if it's a package lying somewhere in the image and Syft picks it up.

I think the alternative option here is to live with duplicates of Pip non-PiPY dependencies =/
Unless you have another idea on how to handle this.

Reply:
What Syft SBOMs will we be merging with this exactly? The one generated from the source directory? From the built image? Both?

Reply:
Both Syft-source and Syft-image SBOMs are already merged in the build pipeline nowadays. So we'll do another merge, and add Cachi2's SBOM there.

I added a Pip few examples [here](https://github.com/brunoapimentel/rhtap-cachi2-sbom-merge), but I still need to update the investigation as I did with Golang.

Reply:
Hmm, theoretically any dependencies unknown to cachi2 should only be able to get there from the base image. I'd say it's fine to drop the Syft duplicates by name in this case

Reply:
Ack. I will make our approach clearer in the docstrings.